eliot/_action.py

```diff
@@ -515,29 +515,37 @@ class WrittenAction(PClass):
     action actually is, and a means of constructing actions that are known to
     be valid.
 
-    @ivar WrittenMessage start_message: The message that started this action.
+    @ivar WrittenMessage start_message: The message that started this
+        action, or C{None} if it is not yet set on the action.
     @ivar WrittenMessage end_message: The message that ends this action. Can be
         C{None} if the action is unfinished.
+    @ivar TaskLevel task_level: The action's task level, e.g. if start
+        message has level C{[2, 3, 1]} it will be
+        C{TaskLevel(level=[2, 3])}.
+    @ivar UUID task_uuid: The UUID of the task to which this action belongs.
     @ivar _children: A L{pmap} from L{TaskLevel} to the L{WrittenAction} and
         L{WrittenMessage} objects that make up this action.
     """
 
-    start_message = field(type=WrittenMessage, mandatory=True)
-    end_message = field(type=optional(WrittenMessage), mandatory=True)
-    # XXX: Actually a map to either WrittenMessage or WrittenAction, but
-    # pyrsistent doesn't support recursive data types:
-    # https://github.com/tobgu/pyrsistent/issues/48
+    start_message = field(type=optional(WrittenMessage), mandatory=True,
+                          initial=None)
+    end_message = field(type=optional(WrittenMessage), mandatory=True,
+                        initial=None)
+    task_level = field(type=TaskLevel, mandatory=True)
+    task_uuid = field(type=unicode, mandatory=True, factory=unicode)
+    # Pyrsistent doesn't support pmap_field with recursive types.
     _children = pmap_field(TaskLevel, object)
 
     @classmethod
-    def from_messages(cls, start_message, children=pvector(), end_message=None):
+    def from_messages(cls, start_message=None, children=pvector(),
+                      end_message=None):
         """
         Create a C{WrittenAction} from C{WrittenMessage}s and other
         C{WrittenAction}s.
 
         @param WrittenMessage start_message: A message that has
             C{ACTION_STATUS_FIELD}, C{ACTION_TYPE_FIELD}, and a C{task_level}
-            that ends in C{1}.
+            that ends in C{1}, or C{None} if unavailable.
         @param children: An iterable of C{WrittenMessage} and C{WrittenAction}
         @param WrittenMessage end_message: A message that has the same
             C{action_type} as this action.
```

Comment:
:grey_question: I noticed you updated the other description of end_message to "An end message that has the same level as this action." I think that is across the board a better description, care to update this one as well?